cmd/cmd.go

```diff
@@ -70,3 +80,327 @@ func validateUserConfig(cfg *viper.Viper) error {
+// allowing people to have multiple parallel versions of an exercise.
+func sanitizeLegacyNumericSuffixFilepath(file, slug string) string {
+	pattern := fmt.Sprintf(`\A.*[/\\]%s-\d*/`, slug)
+	rgxNumericSuffix := regexp.MustCompile(pattern)
+	if rgxNumericSuffix.MatchString(file) {
+		file = string(rgxNumericSuffix.ReplaceAll([]byte(file), []byte("")))
+	}
+	// Rewrite paths submitted with an older, buggy client where the Windows
+	// path is being treated as part of the filename.
+	file = strings.Replace(file, "\\", "/", -1)
+	return filepath.FromSlash(file)
+}
+
+// download is a download from the Exercism API.
+type download struct {
+	*downloadParams
+	*downloadPayload
+	*downloadWriter
+}
+
+// newDownloadFromExercise is a convenience wrapper for creating a new download.
+func newDownloadFromExercise(usrCfg *viper.Viper, exercise workspace.Exercise) (*download, error) {
+	downloadParams, err := newDownloadParamsFromExercise(usrCfg, exercise)
+	if err != nil {
+		return nil, err
+	}
+	return newDownload(downloadParams)
+}
+
+// newDownloadFromFlags is a convenience wrapper for creating a new download.
+func newDownloadFromFlags(usrCfg *viper.Viper, flags *pflag.FlagSet) (*download, error) {
+	downloadParams, err := newDownloadParamsFromFlags(usrCfg, flags)
+	if err != nil {
+		return nil, err
+	}
+	return newDownload(downloadParams)
+}
+
+// newDownload initiates a download by requesting a downloadPayload from the Exercism API.
+func newDownload(params *downloadParams) (*download, error) {
+	if err := params.validate(); err != nil {
+		return nil, err
+	}
+	d := &download{downloadParams: params}
+	d.downloadWriter = &downloadWriter{download: d}
+
+	client, err := api.NewClient(d.usrCfg.GetString("token"), d.usrCfg.GetString("apibaseurl"))
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := client.NewRequest("GET", d.requestURL(), nil)
+	if err != nil {
+		return nil, err
+	}
+	d.buildQuery(req.URL)
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if err := json.NewDecoder(res.Body).Decode(&d.downloadPayload); err != nil {
+		return nil, fmt.Errorf("unable to parse API response - %s", err)
+	}
+
+	if res.StatusCode == http.StatusUnauthorized {
+		siteURL := config.InferSiteURL(d.usrCfg.GetString("apibaseurl"))
+		return nil, fmt.Errorf("unauthorized request. Please run the configure command. You can find your API token at %s/my/settings", siteURL)
+	}
+	if res.StatusCode != http.StatusOK {
+		switch d.Error.Type {
+		case "track_ambiguous":
+			return nil, fmt.Errorf("%s: %s", d.Error.Message, strings.Join(d.Error.PossibleTrackIDs, ", "))
+		default:
+			return nil, errors.New(d.Error.Message)
+		}
+	}
+	return d, d.validate()
+}
+
+func (d *download) requestURL() string {
+	id := "latest"
+	if d.uuid != "" {
+		id = d.uuid
+	}
+	return fmt.Sprintf("%s/solutions/%s", d.usrCfg.GetString("apibaseurl"), id)
+}
+
+func (d *download) buildQuery(url *netURL.URL) {
```

Comment:
AFAICT, this is the only place where `track` or `team` flags are used. It seems that using `--track` or `--team` with `--uuid` renders track or team impotent. If this is indeed an invalid combination of arguments, it seems worth making explicit. An added benefit is clarification of what team/track are used for in `downloadParams`. Rather than commit this now (it might change based on the rest of the review direction), it will be noted here as a potential TODO. Something like: 

```go
func (d *downloadParams) givenUUIDWithTrackOrTeam() error {
	if d.uuid!= "" && (d.team != "" || d.track != "") {
		return errors.New("--team or --track needs --exercise (not --uuid)")
	}
	return nil
}
```